<a href="https://colab.research.google.com/github/raavimanoj016/Portfolio/blob/main/Numpy_FitBit_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Use Case: Fitbit

### Imagine you are a Data Scientist at Fitbit

You've been given a user data to analyse and find some insights which can be shown on the smart watch.

#### But why would we want to analyse the user data for desiging the watch?

These insights from the user data can help business make customer oriented decision for the product design.



#### Lets first look at the data we have gathered

Link: https://drive.google.com/file/d/1Uxwd4H-tfM64giRS1VExMpQXKtBBtuP0/view?usp=sharing

<img src='https://drive.google.com/uc?id=1Uxwd4H-tfM64giRS1VExMpQXKtBBtuP0'>


#### Notice that there are some user features in the data

There are provided as various columns in the data.

#### Every row is called a record or data point


#### What are all the features provided to us? 

- Date
- Step Count
- Mood (Categorical)
- Calories Burned
- Hours of sleep
- Feeling Active (Categorical)


**Using NumPy, we will explore this data to look for some interesting insights - Exploratory Data Analysis.**

#### EDA is all about asking the right questions

#### What kind of questions can we answer using this data?

- How many records and features are there in the dataset?
- What is the **average step count**?
- On which day the **step count was highest/lowest?** 


#### Can we find some deeper insights?

We can probably see how daily activity affects sleep and moood.

We will try finding 
- How daily activity affects mood? 

### Let's first download the dataset

In [ ]:
import numpy as np


In [ ]:
!gdown 1vk1Pu0djiYcrdc85yUXZ_Rqq2oZNcohd


Downloading...
From: https://drive.google.com/uc?id=1vk1Pu0djiYcrdc85yUXZ_Rqq2oZNcohd
To: /content/fit.txt
100% 3.43k/3.43k [00:00<00:00, 4.10MB/s]


In [ ]:
data = np.loadtxt('fit.txt', dtype='str')

In [ ]:
data[:5]

array([['06-10-2017', '5464', 'Neutral', '181', '5', 'Inactive'],
       ['07-10-2017', '6041', 'Sad', '197', '8', 'Inactive'],
       ['08-10-2017', '25', 'Sad', '0', '5', 'Inactive'],
       ['09-10-2017', '5461', 'Sad', '174', '4', 'Inactive'],
       ['10-10-2017', '6915', 'Neutral', '223', '5', 'Active']],
      dtype='<U10')

What's the shape of the data? 

In [ ]:
data.shape

(96, 6)

There are 96 records and each record has 6 features.
These features are:
- Date
- Step count
- Mood
- Calories Burned
- Hours of sleep
- activity status

#### Notice that above array is a homogenous containing all the data as strings

In order to work with strings, categorical data and numerical data, we will have save every feature seperately


We can get some idea on how data is saved.

Lets see whats the first element of `data`

In [ ]:
data[0]

array(['06-10-2017', '5464', 'Neutral', '181', '5', 'Inactive'],
      dtype='<U10')

Hm, this extracts a row not a column


Transpose

In [ ]:
data.T[0]

array(['06-10-2017', '07-10-2017', '08-10-2017', '09-10-2017',
       '10-10-2017', '11-10-2017', '12-10-2017', '13-10-2017',
       '14-10-2017', '15-10-2017', '16-10-2017', '17-10-2017',
       '18-10-2017', '19-10-2017', '20-10-2017', '21-10-2017',
       '22-10-2017', '23-10-2017', '24-10-2017', '25-10-2017',
       '26-10-2017', '27-10-2017', '28-10-2017', '29-10-2017',
       '30-10-2017', '31-10-2017', '01-11-2017', '02-11-2017',
       '03-11-2017', '04-11-2017', '05-11-2017', '06-11-2017',
       '07-11-2017', '08-11-2017', '09-11-2017', '10-11-2017',
       '11-11-2017', '12-11-2017', '13-11-2017', '14-11-2017',
       '15-11-2017', '16-11-2017', '17-11-2017', '18-11-2017',
       '19-11-2017', '20-11-2017', '21-11-2017', '22-11-2017',
       '23-11-2017', '24-11-2017', '25-11-2017', '26-11-2017',
       '27-11-2017', '28-11-2017', '29-11-2017', '30-11-2017',
       '01-12-2017', '02-12-2017', '03-12-2017', '04-12-2017',
       '05-12-2017', '06-12-2017', '07-12-2017', '08-12

Great, we could extract first column

#### Lets extract all the columns and save them in seperate variables

In [ ]:
date, step_count, mood, calories_burned, hours_of_sleep, activity_status = data.T

In [ ]:
step_count

array(['5464', '6041', '25', '5461', '6915', '4545', '4340', '1230', '61',
       '1258', '3148', '4687', '4732', '3519', '1580', '2822', '181',
       '3158', '4383', '3881', '4037', '202', '292', '330', '2209',
       '4550', '4435', '4779', '1831', '2255', '539', '5464', '6041',
       '4068', '4683', '4033', '6314', '614', '3149', '4005', '4880',
       '4136', '705', '570', '269', '4275', '5999', '4421', '6930',
       '5195', '546', '493', '995', '1163', '6676', '3608', '774', '1421',
       '4064', '2725', '5934', '1867', '3721', '2374', '2909', '1648',
       '799', '7102', '3941', '7422', '437', '1231', '1696', '4921',
       '221', '6500', '3575', '4061', '651', '753', '518', '5537', '4108',
       '5376', '3066', '177', '36', '299', '1447', '2599', '702', '133',
       '153', '500', '2127', '2203'], dtype='<U10')

In [ ]:
step_count.dtype

dtype('<U10')

#### Let's convert the data types of these variables

**Step Count**

In [ ]:
step_count = np.array(step_count, dtype = 'int')
step_count.dtype

dtype('int64')

In [ ]:
step_count

array([5464, 6041,   25, 5461, 6915, 4545, 4340, 1230,   61, 1258, 3148,
       4687, 4732, 3519, 1580, 2822,  181, 3158, 4383, 3881, 4037,  202,
        292,  330, 2209, 4550, 4435, 4779, 1831, 2255,  539, 5464, 6041,
       4068, 4683, 4033, 6314,  614, 3149, 4005, 4880, 4136,  705,  570,
        269, 4275, 5999, 4421, 6930, 5195,  546,  493,  995, 1163, 6676,
       3608,  774, 1421, 4064, 2725, 5934, 1867, 3721, 2374, 2909, 1648,
        799, 7102, 3941, 7422,  437, 1231, 1696, 4921,  221, 6500, 3575,
       4061,  651,  753,  518, 5537, 4108, 5376, 3066,  177,   36,  299,
       1447, 2599,  702,  133,  153,  500, 2127, 2203])

**Calories Burned**

In [ ]:
calories_burned = np.array(calories_burned, dtype = 'int')
calories_burned.dtype

dtype('int64')

**Hours of Sleep**

In [ ]:
hours_of_sleep = np.array(hours_of_sleep, dtype = 'int')
hours_of_sleep.dtype

dtype('int64')

**Mood**

`Mood` is a categorical data type. As a name says, categorical data type has two or more categories in it.

Let's check the values of `mood` variable

In [ ]:
mood

array(['Neutral', 'Sad', 'Sad', 'Sad', 'Neutral', 'Sad', 'Sad', 'Sad',
       'Sad', 'Sad', 'Sad', 'Sad', 'Happy', 'Sad', 'Sad', 'Sad', 'Sad',
       'Neutral', 'Neutral', 'Neutral', 'Neutral', 'Neutral', 'Neutral',
       'Happy', 'Neutral', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Neutral', 'Happy', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Neutral',
       'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Neutral', 'Sad', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Happy', 'Happy', 'Sad', 'Neutral', 'Neutral',
       'Sad', 'Sad', 'Neutral', 'Neutral', 'Happy', 'Neutral', 'Neutral',
       'Sad', 'Neutral', 'Sad', 'Neutral', 'Neutral', 'Sad', 'Sad', 'Sad',
       'Sad', 'Happy', 'Neutral', 'Happy', 'Neutral', 'Sad', 'Sad', 'Sad',
       'Neutral', 'Neutral', 'Sad', 'Sad', 'Happy', 'Neutral', 'Neutral',
       'Happy'], dtype='<U10')

In [ ]:
np.unique(mood)

array(['Happy', 'Neutral', 'Sad'], dtype='<U10')

**Activity Status**

In [ ]:
activity_status

array(['Inactive', 'Inactive', 'Inactive', 'Inactive', 'Active',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Active', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Active', 'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Active',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Active', 'Active', 'Active', 'Active', 'Active', 'Active',
       'Active', 'Active', 'Active', 'Inactive', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Inactive', 'Active', 'Active', 'Active',
       'Active', 'Active', 'Active', 'Active', 'Active', 'Active',
       'Active', 'Active', 'Active', 'Inactive', 'Active', 'Active',
       'Inactive', 'Active', 'Active', 'Active', 'Active', 'Active',
       'Inactive', 'Active', 'Active', 'Active', 'Active', 'Inactive',
       'Inactive', 'Inactive', 'Inacti

### Let's try to get some insights from the data.

#### What's the average step count? 

How can we calculate average? => `.mean()`

In [ ]:
step_count.mean()

2935.9375

User moves an average of 2900 steps a day.

#### On which day the step count was highest?

How will be find it? 

First we find the index of maximum step count and use that index to get the date.

How'll we find the index? =>  

Numpy provides a function `np.argmax()` which returns the index of maximum value element.

Similarly, we have a function `np.argmin()` which returns the index of minimum element.

In [ ]:
step_count.argmax()


69

Here 69 is the index of maximum step count element.

In [ ]:
date[step_count.argmax()]

'14-12-2017'

Let's check the calorie burnt on the day

In [ ]:
calories_burned[step_count.argmax()]

243

Not bad! 243 calories. Let's try to get the number of steps on that day as well

In [ ]:
step_count.max()

7422

7k steps!! Sports mode on!

#### Let's try to compare step counts on bad mood days and good mood days

Average step count on Sad mood days

In [ ]:
np.mean(step_count[mood == 'Sad'])

2103.0689655172414

In [ ]:
np.sort(step_count[mood == 'Sad'])

array([  25,   36,   61,  133,  177,  181,  221,  299,  518,  651,  702,
        753,  799, 1230, 1258, 1580, 1648, 1696, 2822, 3148, 3519, 3721,
       4061, 4340, 4545, 4687, 5461, 6041, 6676])

In [ ]:
np.std(step_count[mood == 'Sad'])

2021.2355035376254

Average step count on happy days

In [ ]:
np.mean(step_count[mood == 'Happy'])

3392.725

In [ ]:
np.sort(step_count[mood == 'Happy'])

array([ 153,  269,  330,  493,  539,  546,  614,  705,  774,  995, 1421,
       1831, 1867, 2203, 2255, 2725, 3149, 3608, 4005, 4033, 4064, 4068,
       4136, 4275, 4421, 4435, 4550, 4683, 4732, 4779, 4880, 5195, 5376,
       5464, 5537, 5934, 5999, 6314, 6930, 7422])

Average step count on sad days - 2103.

Average step count on happy days - 3392

There may be relation between mood and step count

#### Let's try to check inverse. Mood when step count was greater/lesser

Mood when step count > 4000

In [ ]:
np.unique(mood[step_count > 4000], return_counts = True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'), array([22,  9,  7]))

Out of 38 days when step count was more than 4000, user was feeling happy on 22 days.

Mood when step count <= 2000

In [ ]:
np.unique(mood[step_count < 2000], return_counts = True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'), array([13,  8, 18]))

Out of 39 days, when step count was less than 2000, user was feeling sad on 18 days.

#### **There may be a correlation between Mood and step count**